documentation/adr/0004-trip-modifications-mqtt-topic-structure.md

```diff
@@ -0,0 +1,179 @@
+
+We know that we'll need these messages to be available if the Transit Data
+service is restarted.
+
+## Decision
+
+We'll use a
+[topic hierarchy](https://www.hivemq.com/blog/mqtt-essentials-part-5-mqtt-topics-best-practices/)
+which allows us to manage individual GTFS-RT messages and update them, and does
+not require that the Transit Data service has some way to persist state between
+restarts.
+
+### Topic Hierarchy
+We'll nest all of our Trip Modifications related data under the
+[MQTT "topic level"](https://www.hivemq.com/blog/mqtt-essentials-part-5-mqtt-topics-best-practices/)
+`trip_modifications`, which will be under our "topic prefix"
+configured in the DevOps repo for each environment.
+
+For Example:
+
+Our `<prefix>`s would be derived from the environment and are described via IaC
+in the devops repo
+```
+skate/prod
+skate/dev
+skate/dev-blue
+skate/dev-green
+```
+And the final combined topic hierarchy would be formed by prepending _one_ of 
+the `<prefix>`s
+```
+<prefix>/trip_modifications/
+```
+For instance,
+```
+skate/dev-blue/trip_modifications/
+```
+
+
+For each new and unique trip modification, we'll generate a unique ID and use
+that as our Trip Modification identifier, named `id`.
+
+> [!NOTE]  
+> The GTFS-RT Trip Modifications spec does not currently have a ID for each modification,
+> so this `ID` _will not_ appear with the Trip Modification Message.
+
+The topic hierarchy is then formed using this ID as the topic level
+```
+<prefix>/trip_modifications/#{id}/
+```
+
+Then, to create or update an associated `trip_modification`, we'll use the topic
+level `trip_modification`
+Creating the topic
+```
+<prefix>/trip_modifications/#{id}/trip_modification
+```
+
+Any associated [`Shape`](https://gtfs.org/realtime/reference/#message-shape)
+Messages will be published to the topic level `shape` in the hierarchy.
+```
+<prefix>/trip_modifications/#{id}/shape
+```
+
+On `Shape` Messages, the
+[`Shape.shape_id`](https://gtfs.org/realtime/reference/#message-shape)
+field will also have a unique ID generated by Skate, which will be
+referenced by the corresponding Trip Modification's
+[`SelectedTrips.shape_id`](https://gtfs.org/realtime/reference/#message-selectedtrips)
+field.
+
+#### Active Trip Modifications State
+Messages pushed to these topics will have the
+[`retain` property](https://www.hivemq.com/blog/mqtt-essentials-part-8-retained-messages/)
+set so that new connections get active Trip Modifications and Shapes published
+to them on connect.
+
+Deactivating a Trip Modification will be done by
+[publishing a zero-byte payload](https://www.hivemq.com/blog/mqtt-essentials-part-8-retained-messages/#heading-how-to-delete-retained-messages-in-mqtt)
+each topic within the `<prefix>/trip_modifications/#{id}/`
+topic hierarchy, which will clear any retained messages and ensure that new
+subscribers are not notified of these now deleted messages and existing
+are made aware of the removal subscribers.
```

Comment:
perhaps this? (I didn't understand the sentence and am taking a stab at copyediting)

```suggestion
subscribers are not notified of these now deleted messages and existing
subscribers are made aware of the removal.
```

Reply:
oops, thanks!